hawkbit-security-core/src/main/java/org/eclipse/hawkbit/security/SecurityContextTenantAware.java

```diff
@@ -59,44 +79,68 @@ public String getCurrentUsername() {
     }
 
     @Override
-    public <T> T runAsTenant(final String tenant, final TenantRunner<T> callable) {
+    public <T> T runAsTenant(final String tenant, final TenantRunner<T> tenantRunner) {
+        return runInContext(buildSystemSecurityContext(tenant), tenantRunner);
+    }
+
+    @Override
+    public <T> T runAsTenantAsUser(final String tenant, final String username, final TenantRunner<T> tenantRunner) {
+        final List<SimpleGrantedAuthority> authorities = runAsSystem(
+                () -> authoritiesResolver.getUserAuthorities(tenant, username).stream().map(SimpleGrantedAuthority::new)
+                        .collect(Collectors.toList()));
+        return runInContext(buildUserSecurityContext(tenant, username, authorities), tenantRunner);
+    }
+
+    private static <T> T runInContext(final SecurityContext context, final TenantRunner<T> tenantRunner) {
         final SecurityContext originalContext = SecurityContextHolder.getContext();
         try {
-            SecurityContextHolder.setContext(buildSecurityContext(tenant));
-            return callable.run();
+            SecurityContextHolder.setContext(context);
+            return tenantRunner.run();
         } finally {
             SecurityContextHolder.setContext(originalContext);
         }
     }
 
-    private static SecurityContext buildSecurityContext(final String tenant) {
+    private static SecurityContext buildSystemSecurityContext(final String tenant) {
+        return buildUserSecurityContext(tenant, SYSTEM_USER, SYSTEM_AUTHORITIES);
+    }
+
+    private static <T> T runAsSystem(final TenantRunner<T> tenantRunner) {
+        final SecurityContext currentContext = SecurityContextHolder.getContext();
+        try {
+            SystemSecurityContext.setSystemContext(currentContext);
```

Comment:
I have a feeling like `SecurityContextTenantAware` and `SystemSecurityContext` belong to the same domain of Security context management and it feels weird to me how they interact with each other. I would definitely not address it in the scope of this PR, just wanted to hear your opinion on the matter?

Reply:
I totally agree. Both classes are heavily intertwined. I think some of the methods were marked static just to avoid the circular reference... ;-)